Control disarmed PWM values per channel through params

sanderux opened the pull request:
```
nsh> param show PWM_MAIN*
Symbols: x = used, + = saved, * = unsaved
x + PWM_MAIN1_DIS [479,522] : 110
x + PWM_MAIN2_DIS [480,523] : 120
x + PWM_MAIN3_DIS [481,524] : 130
x + PWM_MAIN4_DIS [482,525] : 140
x + PWM_MAIN5_DIS [483,526] : 1500
x + PWM_MAIN6_DIS [484,527] : 1500
x + PWM_MAIN7_DIS [485,528] : 900
x + PWM_MAIN8_DIS [486,529] : 900

nsh> pwm info
device: /dev/pwm_output0
channel 1: 110 us (alternative rate: 400 Hz failsafe: 900, disarmed: 110 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 2: 120 us (alternative rate: 400 Hz failsafe: 900, disarmed: 120 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 3: 130 us (alternative rate: 400 Hz failsafe: 900, disarmed: 130 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 4: 140 us (alternative rate: 400 Hz failsafe: 900, disarmed: 140 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 5: 1540 us (default rate: 50 Hz failsafe: 1500, disarmed: 1500 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 6: 1460 us (default rate: 50 Hz failsafe: 1500, disarmed: 1500 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 7: 900 us (default rate: 50 Hz failsafe: 900, disarmed: 900 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 8: 1000 us (default rate: 50 Hz failsafe: 900, disarmed: 900 us, min: 1000 us, max: 2000 us, trim:  0.00)
```


```
nsh> param show PWM_AUX*
Symbols: x = used, + = saved, * = unsaved
x + PWM_AUX1_DIS [457,500] : 210
x + PWM_AUX2_DIS [458,501] : 220
x + PWM_AUX3_DIS [459,502] : 230
x + PWM_AUX4_DIS [460,503] : 240
x + PWM_AUX5_DIS [461,504] : 250
x + PWM_AUX6_DIS [462,505] : 260

nsh> pwm info -d /dev/pwm_output1
device: /dev/pwm_output1
channel 1: 1500 us (default rate: 50 Hz failsafe: 0, disarmed: 210 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 2: 1500 us (default rate: 50 Hz failsafe: 0, disarmed: 220 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 3: 0 us (default rate: 50 Hz failsafe: 0, disarmed: 230 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 4: 0 us (default rate: 50 Hz failsafe: 0, disarmed: 240 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 5: 0 us (default rate: 50 Hz failsafe: 0, disarmed: 250 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 6: 0 us (default rate: 50 Hz failsafe: 0, disarmed: 260 us, min: 1000 us, max: 2000 us, trim:  0.00)

```

LorenzMeier commented on src/modules/sensors/sensor_params.c:
```diff
@@ -3444,6 +3444,261 @@ PARAM_DEFINE_INT32(PWM_MAX, 2000);
 /**
+ * Set the disarmed PWM for the main 1 output
+ *
+ * IMPORTANT: CHANGING THIS PARAMETER REQUIRES A COMPLETE SYSTEM
```
Could we remove this hint as the QGC UI already asks you to reboot?

dagar commented:
Can we try to keep the PWM naming consistent?

PWM_AUX_REV**X**
PWM_AUX_TRIM**X**
PWM_MAIN_REV**X**
PWM_MAIN_TRIM**X**

I wonder if we should have disarmed defaults for motors and servos. That would probably eliminate the need for configuring any of this for the vast majority.

dagar commented on src/modules/sensors/parameters.cpp:
```diff
@@ -232,6 +232,24 @@ int initialize_parameter_handles(ParameterHandles &parameter_handles)
+	(void)param_find("PWM_AUX4_DIS");
+	(void)param_find("PWM_AUX5_DIS");
+	(void)param_find("PWM_AUX6_DIS");
+	
```
Astyle doesn't like the tabs on in these blank lines.

dagar commented on ROMFS/px4fmu_common/init.d/13013_deltaquad:
```diff
@@ -107,12 +107,16 @@ then
+param set PWM_MAIN5_DIS 1500
+param set PWM_MAIN6_DIS 1500
+param set PWM_MAIN7_DIS 900
+param set PWM_MAIN8_DIS 900
```
This is more of a personal preference, but you might want to move this to within the AUTOCNF block. Then users will stick get the default, but could make parameter changes later if needed. Things like trying different ESCs, etc.

sanderux replied:
I have vehicles with clients, they need this forced. when i confirm they have upgraded i will move it to AUTOCNF

sanderux commented:
@LorenzMeier what's going on with semaphore?
```
/opt/gcc/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: src/firmware/nuttx/firmware_nuttx section `.data' will not fit in region `flash'
/opt/gcc/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: region `flash' overflowed by 1676 bytes
```
Is that correct? seems to run fine on FMUV2

dagar commented:
Which compiler version are you using locally?

dagar commented:
Commenting out sdlog2 from px4fmu-v2_default would probably be the easiest.

bkueng commented:
> Commenting out sdlog2 from px4fmu-v2_default would probably be the easiest.

I already did that in https://github.com/PX4/Firmware/pull/7689/commits/c2b32c42b33c975b8b4de89c02e66eb81c17cece, you can just rebase on master.

sanderux commented:
@LorenzMeier should be ready to merge

sanderux commented:
@dagar i addressed all your comments. could you review?

dagar commented:
Setting disarmed to 0 is valid and disables disarmed PWM -
 https://github.com/PX4/Firmware/blob/master/src/systemcmds/pwm/pwm.cpp#L547

I would preserve that and instead use -1 for the special behavior of respecting PWM_DISARMED. Then in the pwm command if -1 is set you can either read PWM_DISARMED (or PWM_AUX_DISARMED) directly (better) or do nothing. https://github.com/PX4/Firmware/blob/master/src/systemcmds/pwm/pwm.cpp#L539

This reduces rc.interface to one liners for each PWM.

    pwm disarmed -c 1 -p p:PWM_MAIN_DIS1
    pwm disarmed -c 2 -p p:PWM_MAIN_DIS2

Are you sure you need `(void)param_find("PWM_MAIN_DIS1");` in sensors parameters.cpp?

sanderux commented:
@dagar i addressed all your comments. could you review?

```
nsh> param show PWM_MAIN_DIS*
Symbols: x = used, + = saved, * = unsaved
x   PWM_MAIN_DIS1 [469,530] : -1
x   PWM_MAIN_DIS2 [470,531] : -1
x   PWM_MAIN_DIS3 [471,532] : -1
x   PWM_MAIN_DIS4 [472,533] : -1
x + PWM_MAIN_DIS5 [473,534] : 500
x   PWM_MAIN_DIS6 [474,535] : -1
x   PWM_MAIN_DIS7 [475,536] : -1
x   PWM_MAIN_DIS8 [476,537] : -1

 947 parameters total, 865 used.
nsh> pwm info
device: /dev/pwm_output0
channel 1: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 2: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 3: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 4: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 5: 1500 us (default rate: 50 Hz failsafe: 1500, disarmed: 500 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 6: 1525 us (default rate: 50 Hz failsafe: 1500, disarmed: 0 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 7: 1475 us (default rate: 50 Hz failsafe: 1500, disarmed: 0 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 8: 1550 us (default rate: 50 Hz failsafe: 1500, disarmed: 0 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel group 0: channels 1 2
channel group 1: channels 5 6 7 8
channel group 2: channels 3 4
```

sanderux commented:
> Are you sure you need (void)param_find("PWM_MAIN_DIS1"); in sensors parameters.cpp?

This is to trigger: https://github.com/PX4/Firmware/blob/master/src/modules/systemlib/param/param.c#L336

dagar commented:
You'll get that from running the pwm command which finds the param. Better yet in setups with only a MAIN the user won't even see the AUX versions because rc.interface won't run that section.

dagar commented:
The metadata at the top of ROMFS/px4fmu_common/init.d/13013_deltaquad is now wrong.

dagar commented on ROMFS/px4fmu_common/init.d/13013_deltaquad:
```diff
@@ -107,12 +107,16 @@ then
+param set PWM_MAIN_DIS7 900
+param set PWM_MAIN_DIS8 900
+
+set MIXER_AUX deltaquad
```
Did you consider using pass.aux.mix here?

dagar commented on ROMFS/px4fmu_common/init.d/rc.interface:
```diff
@@ -227,6 +240,17 @@ then
+		pwm disarmed -c 3 -p p:PWM_AUX_DIS3 -d ${OUTPUT_AUX_DEV}
+		pwm disarmed -c 4 -p p:PWM_AUX_DIS4 -d ${OUTPUT_AUX_DEV}
+		pwm disarmed -c 5 -p p:PWM_AUX_DIS5 -d ${OUTPUT_AUX_DEV}
+		pwm disarmed -c 6 -p p:PWM_AUX_DIS6 -d ${OUTPUT_AUX_DEV}
```
MAIN and AUX are applied inconsistently. For MAIN you only run `pwm disarmed -c X ...` if PWM_OUT is set. For AUX it's done regardless of PWM_AUX_OUT. Is this intentional?

MAIN and AUX aren't consistent here. For MAIN we set this if

sanderux commented:
@dagar i'm done playing this game, this was my last commit to this branch.
feel free to do whatever you like with it

dagar commented:
You don't want correct metadata for your own vehicle? https://dev.px4.io/en/airframes/airframe_reference.html

Handling MAIN and AUX consistently is important so that the correct params sync to the ground control station. The way you have it now only sets all MAIN disarmed params active if PWM_OUT is set at all, but always sets PWM_AUX_DIS*  regardless of PWM_AUX_OUT. Why do it differently? It's easy to do properly.

Adding a redundant mixer wastes space. Things are getting tight on px4fmu-v2_default again. You lose nothing by using what's already in place.

sanderux commented:
The DeltaQuad uses a version of the Dropix FC. That only has 5 AUX channels, so camera control must be run on 1/2 effectively disabling 3/4. 5 would then be used for AUX functionality. 

I'd be willing to make one more pass addressing your comments but at this rate every time i address something you come up with something new. So please tell me if you have stated all your concerns.

sanderux commented:
@LorenzMeier please review

LorenzMeier commented:
Jenkins test this please.

sanderux commented:
@LorenzMeier i rebased to force CI tests

dagar commented:
@sanderux I squashed your commits into separate commits for the new disarm params and then deltaquad changes. I also split the massive sensor_params.c to make it a bit more managable and fixed a few other param errors I noticed.

Tested and looks good. The semaphoreci failure is real and broken in master, I'll fix it and rebase.

sanderux commented:
@dagar Bench tested, works
```
nsh> pwm info
device: /dev/pwm_output0
channel 1: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 2: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 3: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 4: 900 us (alternative rate: 400 Hz failsafe: 900, disarmed: 900 us, min: 1025 us, max: 2000 us, trim:  0.00)
channel 5: 1540 us (default rate: 50 Hz failsafe: 1500, disarmed: 1500 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 6: 1460 us (default rate: 50 Hz failsafe: 1500, disarmed: 1500 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 7: 900 us (default rate: 50 Hz failsafe: 900, disarmed: 900 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel 8: 1000 us (default rate: 50 Hz failsafe: 900, disarmed: 900 us, min: 1000 us, max: 2000 us, trim:  0.00)
channel group 0: channels 1 2
channel group 1: channels 5 6 7 8
channel group 2: channels 3 4
nsh> param show pwm_main_dis*
Symbols: x = used, + = saved, * = unsaved
x   PWM_MAIN_DIS1 [487,530] : -1
x   PWM_MAIN_DIS2 [488,531] : -1
x   PWM_MAIN_DIS3 [489,532] : -1
x   PWM_MAIN_DIS4 [490,533] : -1
x + PWM_MAIN_DIS5 [491,534] : 1500
x + PWM_MAIN_DIS6 [492,535] : 1500
x + PWM_MAIN_DIS7 [493,536] : 900
x + PWM_MAIN_DIS8 [494,537] : 900
```
